File: lib/services/innertube/halal_filter.dart
import 'package:flutter/foundation.dart';
import 'models/innertube_models.dart';

/// Halal content filter for YouTube music
/// 
/// Filters out content that doesn't align with Islamic principles:
/// - Explicit content (marked by YouTube)
/// - Haram keywords (alcohol, drugs, gambling, etc.)
/// - Inappropriate genres/categories
/// 
/// Filter levels:
/// - Strict: Maximum filtering, only clearly halal content passes
/// - Moderate: Balanced filtering, blocks obvious haram content
/// - Light: Minimal filtering, only blocks explicit content
enum FilterLevel { strict, moderate, light }

class HalalFilter {
  FilterLevel _currentLevel;

  // ── Keyword Blacklists ─────────────────────────────────────────────────────

  /// Keywords related to alcohol
  static const List<String> _alcoholKeywords = [
    'alcohol', 'alcoholic', 'beer', 'wine', 'vodka', 'whiskey', 'rum',
    'tequila', 'brandy', 'cocktail', 'drunk', 'drinking', 'bar', 'pub',
    'liquor', 'spirits', 'booze', 'intoxicated', 'hangover', 'brewery',
    'khamr', // Arabic for intoxicant
  ];

  /// Keywords related to drugs/substance abuse
  static const List<String> _drugKeywords = [
    'drug', 'drugs', 'cocaine', 'heroin', 'marijuana', 'weed', 'high',
    'smoking', 'smoke', 'cannabis', 'meth', 'crack', 'pill', 'xanax',
    'oxycontin', 'addict', 'addiction', 'overdose', 'trip', 'tripping',
    'hashish', 'hash', 'narcotic',
  ];

  /// Keywords related to gambling
  static const List<String> _gamblingKeywords = [
    'gamble', 'gambling', 'casino', 'poker', 'bet', 'betting', 'wager',
    'lottery', 'jackpot', 'slot machine', 'roulette', 'blackjack',
    'baccarat', 'dice', 'sportsbook', 'odds', 'horses', 'racing bet',
  ];

  /// Keywords related to explicit sexual content
  static const List<String> _sexualKeywords = [
    'sex', 'sexual', 'nude', 'naked', 'porn', 'pornography', 'explicit',
    'xxx', 'erotic', 'striptease', 'booty', 'ass', 'butt', 'thick',
    'booty shake', 'twerk', 'twink', 'bitch', 'hoe', 'whore', 'slut',
    'pussy', 'dick', 'fuck', 'shit', 'damn', 'hell',
  ];

  /// Keywords related to un-Islamic religious content
  static const List<String> _religiousKeywords = [
    'satan', 'satanic', 'devil worship', 'lucifer', 'occult', 'demon',
    'witch', 'witchcraft', 'spell', 'magic', 'pagan', 'heathen',
    'blasphemy', 'heretic', 'anti christ', 'antichrist',
  ];

  /// Keywords related to violence/harm
  static const List<String> _violenceKeywords = [
    'kill', 'killing', 'murder', 'death', 'die', 'die die', 'shoot',
    'shooting', 'gun', 'guns', 'violence', 'violent', 'blood', 'bloodshed',
    'gang', 'gangsta', 'gangster', 'thug', 'criminal', 'crime', 'rob',
    'steal', 'thief', 'rape', 'abuse', 'torture',
  ];

  /// Music genres often associated with haram content
  static const List<String> _haramGenreKeywords = [
    'gangsta rap', 'drill', 'trap', 'hardcore rap', 'mumble rap',
  ];

  /// Keywords that indicate potentially haram party/club culture
  static const List<String> _partyKeywords = [
    'party', 'partying', 'club', 'clubbing', 'nightclub', 'rave', 'dj',
    'dancehall', 'strip', 'strip club', 'bottle service', 'vip',
    'turnt', 'turn up', 'lit', 'wild', 'crazy', 'after party',
  ];

  /// Keywords related to materialism/excessive wealth promotion
  static const List<String> _materialismKeywords = [
    'billionaire', 'millionaire', 'rich', 'wealth', 'money', 'cash',
    'gold', 'diamond', 'diamonds', 'jewelry', 'jewellery', 'luxury',
    'designer', 'gucci', 'prada', 'louis vuitton', 'rolex', 'bentley',
    'lamborghini', 'ferrari', 'porsche', 'mansion', 'yacht',
  ];

  // ── Halal Positive Indicators ───────────────────────────────────────────────

  /// Keywords that indicate halal/Islamic content (used in strict mode)
  static const List<String> _halalPositiveKeywords = [
    'nasheed', 'islamic', 'quran', 'quran recitation', 'adhkar', 'dua',
    'salah', 'prayer', 'allah', 'muhammad', 'prophet', 'ramadan',
    'eid', 'hajj', 'umrah', 'mosque', 'masjid', 'islamic song',
    'halal', 'deen', 'iman', 'taqwa', 'barakah', 'subhanallah',
    'alhamdulillah', 'allahu akbar', 'mashaallah', 'astaghfirullah',
    'recitation', 'tilawah', 'adhan', 'azan', 'call to prayer',
  ];

  // ── Constructor ────────────────────────────────────────────────────────────

  HalalFilter({FilterLevel level = FilterLevel.moderate}) : _currentLevel = level;

  FilterLevel get currentLevel => _currentLevel;

  void setLevel(FilterLevel level) {
    _currentLevel = level;
    debugPrint('HalalFilter level set to: ${level.name}');
  }

  // ── Filter Methods ─────────────────────────────────────────────────────────

  /// Check if a single YouTube track passes the halal filter
  bool isHalal(YouTubeTrack track) {
    // Always block explicit content
    if (track.isExplicit) return false;

    final textToCheck = _getTextToAnalyze(track);
    final lowerText = textToCheck.toLowerCase();

    switch (_currentLevel) {
      case FilterLevel.strict:
        return _strictFilter(lowerText, track);
      case FilterLevel.moderate:
        return _moderateFilter(lowerText, track);
      case FilterLevel.light:
        return _lightFilter(lowerText, track);
    }
  }

  /// Filter a list of tracks, returning only halal ones
  List<YouTubeTrack> filterTracks(List<YouTubeTrack> tracks) {
    return tracks.where(isHalal).toList();
  }

  /// Get the reason why a track was filtered out (for debugging)
  String getFilterReason(YouTubeTrack track) {
    if (track.isExplicit) return 'Explicit content';

    final textToCheck = _getTextToAnalyze(track);
    final lowerText = textToCheck.toLowerCase();

    // Check all keyword lists
    final checks = [
      ('Alcohol-related content', _alcoholKeywords),
      ('Drug-related content', _drugKeywords),
      ('Gambling-related content', _gamblingKeywords),
      ('Sexual/inappropriate content', _sexualKeywords),
      ('Occult/anti-religious content', _religiousKeywords),
      ('Violence/criminal content', _violenceKeywords),
    ];

    for (final (reason, keywords) in checks) {
      for (final keyword in keywords) {
        if (lowerText.contains(keyword)) {
          return '$reason (matched: "$keyword")';
        }
      }
    }

    if (_currentLevel != FilterLevel.light) {
      final mediumChecks = [
        ('Party/club culture', _partyKeywords),
      ];

      for (final (reason, keywords) in mediumChecks) {
        for (final keyword in keywords) {
          if (lowerText.contains(keyword)) {
            return '$reason (matched: "$keyword")';
          }
        }
      }
    }

    if (_currentLevel == FilterLevel.strict) {
      final strictChecks = [
        ('Materialism/wealth focus', _materialismKeywords),
        ('Haram genre', _haramGenreKeywords),
      ];

      for (final (reason, keywords) in strictChecks) {
        for (final keyword in keywords) {
          if (lowerText.contains(keyword)) {
            return '$reason (matched: "$keyword")';
          }
        }
      }
    }

    return 'Passed filter';
  }

  // ── Filter Level Implementations ───────────────────────────────────────────

  bool _lightFilter(String lowerText, YouTubeTrack track) {
    // Only block explicit and the most offensive content
    return !_containsAny(lowerText, [
      ..._sexualKeywords,
      ..._drugKeywords,
      ..._violenceKeywords.take(5), // Only most violent
    ]);
  }

  bool _moderateFilter(String lowerText, YouTubeTrack track) {
    // Block all haram content categories
    return !_containsAny(lowerText, [
      ..._alcoholKeywords,
      ..._drugKeywords,
      ..._gamblingKeywords,
      ..._sexualKeywords,
      ..._religiousKeywords,
      ..._violenceKeywords,
    ]);
  }

  bool _strictFilter(String lowerText, YouTubeTrack track) {
    // Everything in moderate + party culture + materialism + genre filtering
    final allBlocked = [
      ..._alcoholKeywords,
      ..._drugKeywords,
      ..._gamblingKeywords,
      ..._sexualKeywords,
      ..._religiousKeywords,
      ..._violenceKeywords,
      ..._partyKeywords,
      ..._materialismKeywords,
      ..._haramGenreKeywords,
    ];

    if (_containsAny(lowerText, allBlocked)) return false;

    // In strict mode, prefer content with positive halal indicators
    // If no halal positive keywords found, still allow but log it
    final hasPositiveIndicator = _containsAny(lowerText, _halalPositiveKeywords);
    if (!hasPositiveIndicator) {
      // Don't block, but this could be configured to be stricter
      // For now, we allow content that doesn't match negative keywords
      return true;
    }

    return true;
  }

  // ── Helper Methods ─────────────────────────────────────────────────────────

  /// Get all text from track to analyze
  String _getTextToAnalyze(YouTubeTrack track) {
    final parts = <String>[
      track.title,
      if (track.artist != null) track.artist!,
      if (track.album != null) track.album!,
      if (track.description != null) track.description!,
      if (track.categories != null) track.categories!.join(' '),
    ];
    return parts.join(' ').toLowerCase();
  }

  /// Check if text contains any of the keywords
  bool _containsAny(String text, List<String> keywords) {
    for (final keyword in keywords) {
      if (text.contains(keyword)) {
        return true;
      }
    }
    return false;
  }

  /// Add custom keywords to filter at runtime
  final List<String> _customKeywords = [];

  void addCustomKeyword(String keyword) {
    _customKeywords.add(keyword.toLowerCase());
  }

  void removeCustomKeyword(String keyword) {
    _customKeywords.remove(keyword.toLowerCase());
  }

  void clearCustomKeywords() {
    _customKeywords.clear();
  }

  /// Get all blocked keywords (for settings/info screen)
  Map<String, List<String>> getBlockedKeywords() {
    return {
      'Alcohol': _alcoholKeywords,
      'Drugs': _drugKeywords,
      'Gambling': _gamblingKeywords,
      'Sexual Content': _sexualKeywords,
      'Anti-Religious': _religiousKeywords,
      'Violence': _violenceKeywords,
      'Party/Club': _partyKeywords,
      'Materialism': _materialismKeywords,
      'Haram Genres': _haramGenreKeywords,
      'Custom': _customKeywords,
    };
  }
}

File: halalfilter/src/main/kotlin/com/halalatune/halalfilter/HalalFilter.kt
package com.halalatune.halalfilter

enum class FilterLevel {
    STRICT,
    MODERATE,
    LIGHT
}

data class HalalFilterConfig(
    val level: FilterLevel = FilterLevel.MODERATE,
    val customKeywords: MutableList<String> = mutableListOf()
)

object HalalFilter {
    private val alcoholKeywords = listOf(
        "alcohol", "alcoholic", "beer", "wine", "vodka", "whiskey", "rum",
        "tequila", "brandy", "cocktail", "drunk", "drinking", "bar", "pub",
        "liquor", "spirits", "booze", "intoxicated", "hangover", "brewery",
        "khamr"
    )

    private val drugKeywords = listOf(
        "drug", "drugs", "cocaine", "heroin", "marijuana", "weed", "high",
        "smoking", "smoke", "cannabis", "meth", "crack", "pill", "xanax",
        "oxycontin", "addict", "addiction", "overdose", "trip", "tripping",
        "hashish", "hash", "narcotic"
    )

    private val gamblingKeywords = listOf(
        "gamble", "gambling", "casino", "poker", "bet", "betting", "wager",
        "lottery", "jackpot", "slot machine", "roulette", "blackjack",
        "baccarat", "dice", "sportsbook", "odds", "horses", "racing bet"
    )

    private val sexualKeywords = listOf(
        "sex", "sexual", "nude", "naked", "porn", "pornography", "explicit",
        "xxx", "erotic", "striptease", "booty", "butt", "thick",
        "booty shake", "twerk", "twink", "bitch", "hoe", "whore", "slut",
        "pussy", "dick", "fuck", "shit", "damn", "hell"
    )

    private val religiousKeywords = listOf(
        "satan", "satanic", "devil worship", "lucifer", "occult", "demon",
        "witch", "witchcraft", "spell", "magic", "pagan", "heathen",
        "blasphemy", "heretic", "anti christ", "antichrist"
    )

    private val violenceKeywords = listOf(
        "kill", "killing", "murder", "death", "die", "die die", "shoot",
        "shooting", "gun", "guns", "violence", "violent", "blood", "bloodshed",
        "gang", "gangsta", "gangster", "thug", "criminal", "crime", "rob",
        "steal", "thief", "rape", "abuse", "torture"
    )

    private val haramGenreKeywords = listOf(
        "gangsta rap", "drill", "trap", "hardcore rap", "mumble rap"
    )

    private val partyKeywords = listOf(
        "party", "partying", "club", "clubbing", "nightclub", "rave", "dj",
        "dancehall", "strip", "strip club", "bottle service", "vip",
        "turnt", "turn up", "lit", "wild", "crazy", "after party"
    )

    private val materialismKeywords = listOf(
        "billionaire", "millionaire", "rich", "wealth", "money", "cash",
        "gold", "diamond", "diamonds", "jewelry", "jewellery", "luxury",
        "designer", "gucci", "prada", "louis vuitton", "rolex", "bentley",
        "lamborghini", "ferrari", "porsche", "mansion", "yacht"
    )

    private val halalPositiveKeywords = listOf(
        "nasheed", "islamic", "quran", "quran recitation", "adhkar", "dua",
        "salah", "prayer", "allah", "muhammad", "prophet", "ramadan",
        "eid", "hajj", "umrah", "mosque", "masjid", "islamic song",
        "halal", "deen", "iman", "taqwa", "barakah", "subhanallah",
        "alhamdulillah", "allahu akbar", "mashaallah", "astaghfirullah",
        "recitation", "tilawah", "adhan", "azan", "call to prayer",
        "arabic", "islam", "muslim", "ummah", "jannah", "surah"
    )

    private var config = HalalFilterConfig()

    fun getConfig(): HalalFilterConfig = config

    fun setLevel(level: FilterLevel) {
        config = config.copy(level = level)
    }

    fun addCustomKeyword(keyword: String) {
        config = config.copy(
            customKeywords = config.customKeywords.toMutableList().apply {
                add(keyword.lowercase())
            }
        )
    }

    fun removeCustomKeyword(keyword: String) {
        config = config.copy(
            customKeywords = config.customKeywords.toMutableList().apply {
                remove(keyword.lowercase())
            }
        )
    }

    fun clearCustomKeywords() {
        config = config.copy(customKeywords = mutableListOf())
    }

    fun isHalal(
        title: String,
        artist: String?,
        album: String?,
        description: String?,
        categories: List<String>?,
        isExplicit: Boolean
    ): Boolean {
        if (isExplicit) return false

        val textToCheck = buildString {
            append(title.lowercase())
            artist?.let { append(" ${it.lowercase()}") }
            album?.let { append(" ${it.lowercase()}") }
            description?.let { append(" ${it.lowercase()}") }
            categories?.let { append(" ${it.joinToString(" ").lowercase()}") }
        }

        return when (config.level) {
            FilterLevel.LIGHT -> lightFilter(textToCheck)
            FilterLevel.MODERATE -> moderateFilter(textToCheck)
            FilterLevel.STRICT -> strictFilter(textToCheck)
        }
    }

    private fun lightFilter(text: String): Boolean {
        val keywords = drugKeywords + sexualKeywords + violenceKeywords.take(5)
        return !containsAny(text, keywords)
    }

    private fun moderateFilter(text: String): Boolean {
        val keywords = alcoholKeywords + drugKeywords + gamblingKeywords +
                       sexualKeywords + religiousKeywords + violenceKeywords
        return !containsAny(text, keywords)
    }

    private fun strictFilter(text: String): Boolean {
        val allBlocked = alcoholKeywords + drugKeywords + gamblingKeywords +
                         sexualKeywords + religiousKeywords + violenceKeywords +
                         partyKeywords + materialismKeywords + haramGenreKeywords +
                         config.customKeywords

        if (containsAny(text, allBlocked)) return false
        return true
    }

    private fun containsAny(text: String, keywords: List<String>): Boolean {
        return keywords.any { text.contains(it) }
    }

    fun getFilterReason(
        title: String,
        artist: String?,
        album: String?,
        description: String?,
        categories: List<String>?,
        isExplicit: Boolean
    ): String {
        if (isExplicit) return "Explicit content"

        val textToCheck = buildString {
            append(title.lowercase())
            artist?.let { append(" ${it.lowercase()}") }
            album?.let { append(" ${it.lowercase()}") }
            description?.let { append(" ${it.lowercase()}") }
            categories?.let { append(" ${it.joinToString(" ").lowercase()}") }
        }

        val checks = listOf(
            "Alcohol-related content" to alcoholKeywords,
            "Drug-related content" to drugKeywords,
            "Gambling-related content" to gamblingKeywords,
            "Sexual/inappropriate content" to sexualKeywords,
            "Occult/anti-religious content" to religiousKeywords,
            "Violence/criminal content" to violenceKeywords
        )

        for ((reason, keywords) in checks) {
            for (keyword in keywords) {
                if (textToCheck.contains(keyword)) {
                    return "$reason (matched: \"$keyword\")"
                }
            }
        }

        if (config.level != FilterLevel.LIGHT) {
            for ((reason, keywords) in listOf("Party/club culture" to partyKeywords)) {
                for (keyword in keywords) {
                    if (textToCheck.contains(keyword)) {
                        return "$reason (matched: \"$keyword\")"
                    }
                }
            }
        }

        if (config.level == FilterLevel.STRICT) {
            for ((reason, keywords) in listOf(
                "Materialism/wealth focus" to materialismKeywords,
                "Haram genre" to haramGenreKeywords
            )) {
                for (keyword in keywords) {
                    if (textToCheck.contains(keyword)) {
                        return "$reason (matched: \"$keyword\")"
                    }
                }
            }
        }

        return "Passed filter"
    }

    fun getBlockedKeywords(): Map<String, List<String>> = mapOf(
        "Alcohol" to alcoholKeywords,
        "Drugs" to drugKeywords,
        "Gambling" to gamblingKeywords,
        "Sexual Content" to sexualKeywords,
        "Anti-Religious" to religiousKeywords,
        "Violence" to violenceKeywords,
        "Party/Club" to partyKeywords,
        "Materialism" to materialismKeywords,
        "Haram Genres" to haramGenreKeywords,
        "Custom" to config.customKeywords
    )

    fun reset() {
        config = HalalFilterConfig()
    }
}
